syntax = "proto3";
package vdss;

// ── Core types ──────────────────────────────────────────────────────────────

message VectorIdentifier {
  oneof id {
    uint64 u64_id = 1;
    string uuid   = 2;
  }
}

message Vector {
  repeated float data = 1;
  uint32 dimension    = 2;
}

message Payload {
  string json = 1;
}

message Status {
  int32  code    = 1;
  string message = 2;
}

// ── UpsertVector ─────────────────────────────────────────────────────────────

message UpsertVectorRequest {
  string          collection_name = 1;
  VectorIdentifier vector_id      = 2;
  Vector           vector         = 3;
  Payload          payload        = 4;
}

message UpsertVectorResponse {
  Status status = 1;
}

// ── Search ───────────────────────────────────────────────────────────────────

message SearchRequest {
  string collection_name  = 1;
  Vector query            = 2;
  uint32 top_k            = 3;
  optional string filter_json = 4;
  bool with_vector        = 5;
  bool with_payload       = 6;
}

message SearchResult {
  VectorIdentifier  id      = 1;
  float             score   = 2;
  optional Vector   vector  = 3;
  optional Payload  payload = 4;
}

message SearchResponse {
  Status              status  = 1;
  repeated SearchResult results = 2;
}

// ── DeleteVector ─────────────────────────────────────────────────────────────

message DeleteVectorRequest {
  string           collection_name = 1;
  VectorIdentifier vector_id       = 2;
}

message DeleteVectorResponse {
  Status status = 1;
}

// ── CreateCollection ─────────────────────────────────────────────────────────

enum IndexDriver {
  FAISS = 0;
}

enum IndexAlgorithm {
  HNSW = 0;
}

enum StorageType {
  BTRIEVE_FILE = 0;
  MEM          = 1;
  BTRIEVE_SPACE = 2;
}

enum DistanceMetric {
  COSINE     = 0;
  EUCLIDEAN  = 1;
  DOT        = 2;
}

message HnswConfig {
  uint32 m               = 1;
  uint32 ef_construction = 2;
  uint32 ef_search       = 3;
}

message CollectionConfig {
  IndexDriver              index_driver     = 1;
  IndexAlgorithm           index_algorithm  = 2;
  StorageType              storage_type     = 3;
  uint32                   dimension        = 4;
  DistanceMetric           distance_metric  = 5;
  string                   config_json      = 6;
  optional HnswConfig      hnsw_config      = 7;
}

message CreateCollectionRequest {
  string           collection_name = 1;
  CollectionConfig config          = 2;
}

message CreateCollectionResponse {
  Status status = 1;
}

// ── Service ──────────────────────────────────────────────────────────────────

message FlushRequest {
  string collection_name = 1;
}

message FlushResponse {
  Status status = 1;
}

service VDSSService {
  rpc UpsertVector     ( UpsertVectorRequest     ) returns ( UpsertVectorResponse     );
  rpc Search           ( SearchRequest           ) returns ( SearchResponse           );
  rpc DeleteVector     ( DeleteVectorRequest     ) returns ( DeleteVectorResponse     );
  rpc CreateCollection ( CreateCollectionRequest ) returns ( CreateCollectionResponse );
  rpc Flush ( FlushRequest ) returns ( FlushResponse );
}
